README.md
# colossal-squuid

[![CI](https://github.com/yetanalytics/colossal-squuid/actions/workflows/main.yml/badge.svg)](https://github.com/yetanalytics/colossal-squuid/actions/workflows/main.yml) [![Clojars Project](https://img.shields.io/clojars/v/com.yetanalytics/colossal-squuid.svg)](https://clojars.org/com.yetanalytics/colossal-squuid) [![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-2.1-5e0b73.svg)](CODE_OF_CONDUCT.md)


Library for generating Sequential UUIDs, or SQUUIDs.

## Overview

A SQUUID is a Universally Unique Identifier, or UUID, whose value increases strictly monotonically over time. A SQUUID generated later will always have a higher value, both lexicographically and in terms of the underlying bits, than one generated earlier. This is in contrast to regular UUIDs (specifically version 4 UUIDs) that are completely random. However, it is also useful for generated SQUUIDs to maintain some degree of randomness to preserve uniqueness and reduce collision, rather than being completely one-to-one with a particular timestamp.

## Installation

If using deps.edn, add the following line to your `:deps` map:
```clojure
com.yetanalytics/colossal-squuid {:mvn/version "0.1.5"}
```
See the [Clojars page](https://clojars.org/com.yetanalytics/colossal-squuid) for how to install via Leiningen or other methods.

**Note:** By default colossal-squuid will bring in the Clojure and ClojureScript libraries as transitive dependencies. If you wish to exclude these from your project (e.g. because it is clj or cljs-only), you can use the `:exclusions` keyword (which works for both [deps.edn](https://simonrobson.net/2019/04/16/clojure-deps-with-exclusions.html) and [Leiningen](https://github.com/technomancy/leiningen/blob/master/sample.project.clj#L55)):
```clojure
:exclusions [org.clojure/clojure org.clojure/clojurescript]
```

For earlier releases, see the [Clojars page](https://clojars.org/com.yetanalytics/colossal-squuid), as well as the [changelog](CHANGELOG.md) for information about each release.

## API

Four functions are provided in the `com.yetanalytics.squuid` namespace:
- `generate-squuid` generates a SQUUID based off of a random base UUID and a timestamp representing the current time.
- `generate-squuid*`, which returns a map containing the base UUID, the timestamp, and the SQUUID.
- `time->uuid` takes a timestamp and creates a SQUUID with a fixed (not random)  base UUID portion.
- `uuid->time` takes a SQUUID and returns its corresponding timestamp.

**Note:** `generate-squuid*` and `uuid->time` return timestamps as `java.time.Instant` instances in Clojure, `#inst` in ClojureScript.

```clojure
(generate-squuid)
;; => #uuid "017de28f-5801-8c62-9ce9-cef70883794a"

(generate-squuid*)
;; => {:timestamp #inst "2021-12-22T14:33:04.769000000-00:00"
;;     :base-uuid #uuid "85335e1f-9c1f-4c62-9ce9-cef70883794a"
;;     :squuid    #uuid "017de28f-5801-8c62-9ce9-cef70883794a"}

(time->uuid #inst "2021-12-22T14:33:04.769000000-00:00")
;; => #uuid "017de28f-5801-8fff-8fff-ffffffffffff"

(uuid->time #uuid "017de28f-5801-8fff-8fff-ffffffffffff")
;; => #inst "2021-12-22T14:33:04.769000000-00:00"
```

## Implementation

Our solution is to generate SQUUIDs where the first 48 bits are timestamp-based, while the remaining 80 bits are derived from a v4 base UUID. Abiding by RFC 4122, there are 6 reserved bits in a v4 UUID: 4 for the version (set at `0100`) and 2 for the variant (set at `11`). This means that there are 74 remaining random bits, which allows for about 18.9 sextillion random segments.

The timestamp is coerced to millisecond resolution. Specifically, it is the number of milliseconds since the start of the UNIX epoch on January 1, 1970. Due to the 48 bit maximum on the timestamp, the latest time supported is August 2, 10889; any millisecond-resolution timestamp generated after this date will have more than 48 bits.

If two SQUUIDs are generated in the same milliseconds, then instead of using completely different base UUIDs, the earlier SQUUID will be incremented by 1 to create the later SQUUID, ensuring strict monotonicity. In the very unlikely case the SQUUID cannot be incremented, an exception will be thrown.

The generated SQUUIDs have a version number of 8, as that is the version suggested by the draft RFC on SQUUIDs (see "Background").

A graphical representation of the generated SQUUIDs is as follows:
```
|-timestamp-| |----base v4 UUID----|
xxxxxxxx-xxxx-Mxxx-Nxxx-xxxxxxxxxxxx
```
Where `M` is the version (always set to `8`) and `N` is the variant (which, since only the two most significant bits are fixed, can range from `8` to `B`).

## Background

- The original approach of generating a 48-bit timestamp and merging it into a v4 UUID is taken from the Laravel PHP library's `orderedUuid` function: https://itnext.io/laravel-the-mysterious-ordered-uuid-29e7500b4f8
- The idea of incrementing the least significant bit on a timestamp collision is taken from the ULID specification: https://github.com/ulid/spec
- The draft RFC for v8 UUIDs: https://datatracker.ietf.org/doc/html/draft-peabody-dispatch-new-uuid-format

## Contribution

Before contributing to this project, please read the [Contribution Guidelines](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

## License

Copyright © 2021-2022 Yet Analytics, Inc.

colossal-squuid is licensed under the Apache License, Version 2.0. See [LICENSE](https://github.com/yetanalytics/colossal-squuid/blob/main/LICENSE) for the full license text


CHANGELOG.md
# Change Log

## 0.1.5
- Update GitHub CI and CD to remove deprecation warnings.

## 0.1.4

- Update README by adding an API section. ([#16](https://github.com/yetanalytics/colossal-squuid/pull/16))
- Update CD workflow such that Clojars deployment occurs in a separate GitHub Action. ([#17](https://github.com/yetanalytics/colossal-squuid/pull/17))
- Add changelog, contributing guidelines, and code of conduct. ([#19](https://github.com/yetanalytics/colossal-squuid/pull/19))
- Add the `uuid->time` API function to extract the timestamp from a SQUUID. ([#18](https://github.com/yetanalytics/colossal-squuid/pull/18))

## 0.1.3

- Fix bug from 0.1.2 where JAR file is not compiled with the correct files. ([#15](https://github.com/yetanalytics/colossal-squuid/pull/15))

## 0.1.2

- Incorporate automatic Clojars deployment in CD workflow. ([#14](https://github.com/yetanalytics/colossal-squuid/pull/14), with ideas incorporated from [#13](https://github.com/yetanalytics/colossal-squuid/pull/13))

## 0.1.1

- Fix [issue](https://github.com/yetanalytics/colossal-squuid/issues/10) where strict monotonicity was not enforced on multiple threads. ([#11](https://github.com/yetanalytics/colossal-squuid/pull/11))

## 0.1.0

Initial release!


src/test/com/yetanalytics/squuid_test.clj
(ns com.yetanalytics.squuid-test
  (:require [clojure.test :refer [deftest testing is]]
            [com.yetanalytics.squuid :as squuid]
            [com.yetanalytics.squuid.uuid :refer [before? equals?]]
            [com.yetanalytics.squuid.test-support :refer [check-validate]]))

(deftest squuid-test
  (testing "squuid gentests"
    (is (nil? (check-validate `squuid/generate-squuid*)))
    (is (nil? (check-validate `squuid/generate-squuid)))
    (is (nil? (check-validate `squuid/time->uuid))))
  (testing "squuid monotonicity"
    (is
     (loop [squuid-seq  (repeatedly 10000 squuid/generate-squuid)
            squuid-seq' (rest squuid-seq)]
       (if (empty? squuid-seq')
         true
         (let [prev-squuid (first squuid-seq)
               next-squuid (first squuid-seq')]
           (if (before? prev-squuid next-squuid)
             (recur squuid-seq' (rest squuid-seq'))
             false))))))
  (testing "squuid monotonicity (sort)"
    (let [squuid-seq  (repeatedly 1000 squuid/generate-squuid)
          squuid-seq' (sort (fn [u1 u2] (.compareTo u1 u2)) squuid-seq)]
      (is (every? (fn [[u1 u2]] (equals? u1 u2))
                  (map (fn [u1 u2] [u1 u2]) squuid-seq squuid-seq'))))))


src/test/com/yetanalytics/squuid/test_support.clj
(ns com.yetanalytics.squuid.test-support
  (:require [clojure.spec.test.alpha :as stest]))

;; Copied from training-commons.xapi.statement-gen-test
(defn check-validate
  "Given the function name `fname`, returns `nil` if its generative
   tests passes, the erroneous result otherwise. If `num-tests` is
   not provided, runs 50 tests by default."
  ([fname]
   (check-validate fname 50))
  ([fname num-tests]
   (let [opts {:clojure.spec.test.check/opts
               {:num-tests num-tests
                :seed      (rand-int Integer/MAX_VALUE)}}
         res (stest/check fname opts)]
     (when-not (true? (-> res first :clojure.spec.test.check/ret :pass?))
       res))))


src/main/com/yetanalytics/squuid/time.clj
(ns com.yetanalytics.squuid.time
  (:import [java.time Instant]))

(set! *warn-on-reflection* true)

(def ^:private max-time-emsg
  (str "Cannot generate SQUUID past August 2, 10889."
       " The java.time.Instant timestamp would have exceeded 48 bits."))

(defn- assert-valid-time
  [t]
  (when-not (< t 0x0000FFFFFFFFFFFF)
    (throw (ex-info max-time-emsg
                    {:type ::exceeded-max-time
                     :time t}))))

(defn ^Instant zero-time
  "Return the Instant instance representing the beginning of the UNIX epoch."
  []
  (Instant/EPOCH))

(defn ^Instant current-time
  "Return the Instant instance for the current system time. Throws an exception
   if the underlying long exceeds 48 bits."
  []
  (let [curr-time (System/currentTimeMillis)]
    (assert-valid-time curr-time)
    (Instant/ofEpochMilli curr-time)))

(defn before?
  "Does `inst1` come before `inst2`?"
  [^Instant inst1 ^Instant inst2]
  (.isBefore inst1 inst2))

(defn after?
  "Does `inst1` come after `inst2`?"
  [^Instant inst1 ^Instant inst2]
  (.isAfter inst1 inst2))

(defn time->millis
  "Given an Instant instance `inst`, convert it into the number of milliseconds
   since the start of the UNIX epoch."
  [^Instant inst]
  (.toEpochMilli inst))

(set! *warn-on-reflection* false)


src/main/com/yetanalytics/squuid/uuid.clj
(ns com.yetanalytics.squuid.uuid
  (:require [com.yetanalytics.squuid.time :as t])
  (:import [java.util UUID]))

(set! *warn-on-reflection* true)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Helpers + Java interop wrappers
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn ^UUID zero-uuid
  "Generate a v0 UUID of all zeroes."
  []
  #uuid "00000000-0000-0000-0000-000000000000")

(defn ^UUID rand-uuid
  "Generate a random v4 UUID."
  []
  (UUID/randomUUID))

(defn ^Long uuid-msb
  "Get the most significant bytes of a UUID."
  [^UUID uuid]
  (.getMostSignificantBits uuid))

(defn ^Long uuid-lsb
  "Get the least significant bytes of a UUID."
  [^UUID uuid]
  (.getLeastSignificantBits uuid))

(defn bytes->uuid
  "Return a UUID given the most and least significant bytes."
  [^Long msb ^Long lsb]
  (UUID. msb lsb))

;; uuid<, uuid=, and uuid> from clj-uuid

(defn before?
  "Does `uuid1?` come before `uuid2`?"
  [^UUID uuid1 ^UUID uuid2]
  (neg-int? (.compareTo uuid1 uuid2)))

(defn equals?
  "Does `uuid1?` equal `uuid2`?"
  [^UUID uuid1 ^UUID uuid2]
  (zero? (.compareTo uuid1 uuid2)))

(defn after?
  "Does `uuid1?` come after `uuid2`?"
  [^UUID uuid1 ^UUID uuid2]
  (pos-int? (.compareTo uuid1 uuid2)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Public functions
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(def ^{:private true :const true} bit-mask-12
  (unchecked-long 0x0000000000000FFF))
(def ^{:private true :const true} bit-mask-16
  (unchecked-long 0x000000000000FFFF))
(def ^{:private true :const true} bit-mask-61
  (unchecked-long 0x1FFFFFFFFFFFFFFF))

(defn inc-uuid
  "Increment the UUID by one bit. Throws an exception if there are no available
   bits left to increment."
  [uuid]
  (let [uuid-msb (uuid-msb uuid)
        uuid-lsb (uuid-lsb uuid)]
    (cond
      ;; least significant bits not maxed out
      (not (zero? (bit-and-not bit-mask-61 uuid-lsb)))
      (bytes->uuid uuid-msb (inc uuid-lsb))
      ;; most significant bits not maxed out
      (not (zero? (bit-and-not bit-mask-12 uuid-msb)))
      (bytes->uuid (inc uuid-msb) uuid-lsb)
      ;; oh no
      :else
      (throw (ex-info (format "Cannot increment UUID %s any further."
                              uuid)
                      {:type ::exceeded-max-uuid-node
                       :uuid uuid})))))

(defn make-squuid
  "Make a new v8 sequential UUID. Uses `uuid` as the base UUID if provided;
   otherwise uses a random v4 UUID as the base."
  ([ts]
   (make-squuid ts (rand-uuid)))
  ([ts uuid]
   (let [;; Base UUID
         uuid-msb  (uuid-msb uuid)
         uuid-lsb  (uuid-lsb uuid)
        ;; Timestamp
         ts-long   (t/time->millis ts)
        ;; Putting it all together (and change version from v4 to v8)
         uuid-msb' (-> (bit-or (bit-shift-left ts-long 16)
                               (bit-and bit-mask-16 uuid-msb))
                       (bit-clear 14)
                       (bit-set 15))
         squuid    (bytes->uuid uuid-msb' uuid-lsb)]
     {:base-uuid uuid
      :squuid    squuid})))

(defn compare-uuid
  "Return:
   - (< uuid1 uuid2): -1
   - (= uuid1 uuid2): 0
   - (> uuid1 uuid2): 1"
  [uuid1 uuid2]
  (.compareTo uuid1 uuid2))

(set! *warn-on-reflection* false)


src/main/com/yetanalytics/squuid.clj
(ns com.yetanalytics.squuid
  (:require [clojure.spec.alpha :as s]
            [com.yetanalytics.squuid.uuid :as u]
            [com.yetanalytics.squuid.time :as t])
  (:import [java.util Date]))

;; This library generates sequential UUIDs, or SQUUIDs, based on the draft RFC
;; for v8 UUIDS:
;; https://datatracker.ietf.org/doc/html/draft-peabody-dispatch-new-uuid-format

;; The original approach of generating a 48-bit timestamp and merging it into
;; a v4 UUID is taken from the Laravel PHP library's orderedUuid function:
;; https://itnext.io/laravel-the-mysterious-ordered-uuid-29e7500b4f8

;; The idea of incrementing the least significant bit on a timestamp collision
;; is taken from the ULID specification:
;; https://github.com/ulid/spec

(s/def ::base-uuid uuid?)
(s/def ::squuid uuid?)
(s/def ::timestamp inst?)

;; The atom is private so that only generate-squuid(*) can mutate it.
;; Note that merging Instant/EPOCH with v0 UUID returns the v0 UUID again.
(def ^:private current-time-atom
    (atom {:timestamp (t/zero-time)
           :base-uuid (u/zero-uuid)
           :squuid    (u/zero-uuid)}))

(s/fdef generate-squuid*
  :args (s/cat)
  :ret (s/keys :req-un [::base-uuid ::timestamp ::squuid]))

(defn generate-squuid*
  "Return a map containing the following:
   :squuid     The v8 sequential UUID made up of a base UUID and timestamp.
   :base-uuid  The base v4 UUID that provides the lower 80 bits.
   :timestamp  The timestamp that provides the higher 48 bits.
   
   The sequential UUIDs have 7 reserved bits from the RFC 4122 standard;
   4 for the UUID version and 2 for the UUID variant. This leaves 74 random
   bits, allowing for about 18.9 sextillion random segments.
   
   The timestamp is coerced to millisecond resolution. Due to the 48 bit
   maximum on the timestamp, the latest time supported is February 11, 10332.
   
   In case that this function is called multiple times in the same millisecond,
   subsequent SQUUIDs are created by incrementing the base UUID and thus the
   random segment of the SQUUID. An exception is thrown in the unlikely case
   where all 72 (non-variant) random bits are 1s and incrementing can no
   longer occur."
  []
  (let [ts (t/current-time)
        {:keys [timestamp]} @current-time-atom]
    (if-not (t/after? ts timestamp)
      ;; Timestamp clash - increment UUIDs
      (swap! current-time-atom (fn [m]
                                 (-> m
                                     (update :base-uuid u/inc-uuid)
                                     (update :squuid u/inc-uuid))))
      ;; No timestamp clash - make new UUIDs
      (swap! current-time-atom (fn [m]
                                 (-> m
                                     (assoc :timestamp ts)
                                     (merge (u/make-squuid ts))))))))

(s/fdef generate-squuid
  :args (s/cat)
  :ret ::squuid)

(defn generate-squuid
  "Return a new v8 sequential UUID, or SQUUID. The most significant 48 bits
   are created from a timestamp representing the current time, which always
   increments, enforcing sequentialness. See `generate-squuid*` for more
   details."
  []
  (:squuid (generate-squuid*)))

(s/fdef time->uuid
  :args (s/cat :ts inst?)
  :ret ::squuid)

(defn time->uuid
  "Convert a timestamp to a UUID. The upper 48 bits represent
   the timestamp, while the lower 80 bits are `8FFF-8FFF-FFFFFFFFFFFF`."
  [ts]
  (let [ts (if (instance? Date ts) (.toInstant ts) ts)]
    (:squuid
     (u/make-squuid ts #uuid "00000000-0000-4FFF-8FFF-FFFFFFFFFFFF"))))
